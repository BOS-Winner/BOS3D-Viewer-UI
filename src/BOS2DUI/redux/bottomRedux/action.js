import * as actionType from "./actionType";

export function changeMode(mode) {
  return {
    type: actionType.CHANGE_MODE,
    mode,
  };
}

export function showSetting(show) {
  return {
    type: actionType.SHOW_SETTING,
    show,
  };
}

export function showCptInfo(show) {
  return {
    type: actionType.SHOW_CPTINFO,
    show,
  };
}

export function showLayerSwitcher(show) {
  return {
    type: actionType.SHOW_LAYER_SWITCHER,
    show,
  };
}

export function updateLayerList(layers) {
  return {
    type: actionType.UPDATE_LAYER_LIST,
    layers,
  };
}

/**
 * 打开或隐藏批注面板
 * @param {boolean} show - 是否显示
 * @return {{show: boolean, type: string}}
 */
export function showAnnotationList(show) {
  return {
    type: actionType.SHOW_ANNOTATION_LIST,
    show,
  };
}

/**
 * 打开或隐藏批注编辑器
 * @param {boolean} show - 是否显示
 * @param {object} [data] - 是否使用指定预设数据
 * @return {{show: boolean, type: string, data:object}}
 */
export function showAnnotationEditor(show, data) {
  return {
    type: actionType.SHOW_ANNOTATION_EDITOR,
    show,
    data,
  };
}
