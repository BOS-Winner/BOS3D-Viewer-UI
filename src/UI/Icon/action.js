import * as actionType from "./actionType";

/**
 * 改变当前UI模式
 * @param {string} mode - 模式名称
 * @return {{mode: string, type: string}}
 */
export function changeMode(mode, exitMode = "") {
  return {
    type: actionType.CHANGE_MODE,
    mode,
    exitMode,
  };
}

/**
 * 更新操作列表
 * @param {object} obj - 操作数据，包括key和操作名称，其他必需数据
 * @param {string} obj.type - 操作类型（方法）: hide, wireframe, isolate, colorful
 * @param {string} obj.name - 名称
 * @param {string[]} obj.keys - 操作的构件key数组
 * @param {number} [obj.color] - 构件改变的颜色
 * @return {{item: object, type: string}}
 */
export function updateList(obj) {
  return {
    type: actionType.UPDATE_LIST,
    item: obj
  };
}

/**
 * 移动操作列表尾指针
 * @param {number} pointer - 新的指针位置
 * @return {{pointer: number, type: string}}
 */
export function movePointer(pointer) {
  return {
    type: actionType.MOVE_POINTER,
    pointer
  };
}

/**
 * 重设操作历史
 * @return {{type: string}}
 */
export function resetHistory() {
  return {
    type: actionType.RESET_HISTORY
  };
}

/**
 * 修改鼠标图标
 * @param mouseIcon
 * @return {{mouseIcon: *, type: string}}
 */
export function changeMouseIcon(mouseIcon) {
  return {
    type: actionType.CHANGE_MOUSE_ICON,
    mouseIcon,
  };
}

/**
 * 构件信息是否显示
 * @param visible
 * @return {{visible: boolean, type: string}}
 */
export function setComponentInfoVisible(visible) {
  return {
    type: actionType.COMPONENT_INFO_VISIBLE,
    visible
  };
}

/**
 * 部件族信息是否显示
 * @param {object} payload 部件族信息显示的参数
 */
export function setFamilyInfoVisible(payload) {
  return {
    type: actionType.FAMILY_INFO_VISIBLE,
    payload
  };
}
