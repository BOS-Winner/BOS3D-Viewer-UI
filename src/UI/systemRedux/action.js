import * as actionType from "./actionType";

/**
 * 保存模型信息到reducer
 */
export function updateModelDetail(payload) {
  return {
    type: actionType.UPDATE_MODEL_DETAIL,
    payload
  };
}
