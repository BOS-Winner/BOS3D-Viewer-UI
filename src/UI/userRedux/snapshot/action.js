import * as actionType from "./actionType";

export function add(snapshot) {
  return {
    type: actionType.ADD,
    snapshot
  };
}

export function remove(code) {
  return {
    type: actionType.REMOVE,
    code
  };
}

export function update(code, snapshot) {
  return {
    type: actionType.UPDATE,
    snapshot,
    code
  };
}

export function clear() {
  return {
    type: actionType.CLEAR
  };
}
