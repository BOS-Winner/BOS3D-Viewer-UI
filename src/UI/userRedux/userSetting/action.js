import * as actionType from "./actionType";

export function changeToolbarState(name, enable) {
  return {
    type: actionType.CHANGE_TOOLBAR_STATE,
    name,
    enable,
  };
}

export function setToolbarState(toolState) {
  return {
    type: actionType.SET_TOOLBAR_STATE,
    toolState,
  };
}

export function setCustomToolbarState(toolState) {
  return {
    type: actionType.SET_CUSTOMTOOLSTATE,
    customToolState: toolState,
  };
}

export function changeDisplaySetting(name, value) {
  return {
    type: actionType.CHANGE_DISPLAY_SETTING,
    name,
    value,
  };
}

export function changeCameraSetting(name, value) {
  return {
    type: actionType.CHANGE_CAMERA_SETTING,
    name,
    value
  };
}

export function restoreSetting(settingType) {
  return {
    type: actionType.RESTORE_SETTING,
    settingType,
  };
}

export function changeModelSetting(modelKey, name, value) {
  return {
    type: actionType.CHANGE_MODEL_SETTING,
    modelKey,
    name,
    value,
  };
}

export function setSetting(setting) {
  return {
    type: actionType.SET_SETTING,
    setting,
  };
}

export function changeBestView(visible = false) {
  return {
    type: actionType.CHANGE_BEST_VIEW,
    visible,
  };
}
