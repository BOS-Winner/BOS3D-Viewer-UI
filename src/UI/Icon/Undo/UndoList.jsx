import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import _ from "lodash-es";
import { movePointer } from "../action";
import style from "./style.less";
import Modal from "../../Base/Modal";
import { AntdIcon } from '../../utils/utils';
import Empty from '../../Base/Empty';
import { DEFAULT_MODAL_PLACE } from "../../constant";

class UndoList extends React.PureComponent {
  constructor(props) {
    super(props);
    const viewer3D = this.props.viewer;
    this.viewerFunc = {
      undo: {
        hide: viewer3D.showComponentsByKey.bind(viewer3D),
        isolate: keyObj => {
          viewer3D.resetScene({
            visible: true
          });
          if (keyObj.isolate.length > 0) {
            viewer3D.isolateComponentsByKey(keyObj.isolate);
          }
          viewer3D.hideComponentsByKey(keyObj.hide);
        },
        colorful: viewer3D.closeColorfulComponentsByKey.bind(viewer3D),
        wireframe: viewer3D.closeWireFrameComponentsByKey.bind(viewer3D),
        transparent: viewer3D.clearTransparentList.bind(viewer3D), // 显示所有构件
      },
      redo: {
        hide: viewer3D.hideComponentsByKey.bind(viewer3D),
        isolate: viewer3D.isolateComponentsByKey.bind(viewer3D),
        colorful: viewer3D.colorfulComponentsByKey.bind(viewer3D),
        wireframe: viewer3D.wireFrameComponentsByKey.bind(viewer3D),
        transparent: viewer3D.transparentOtherComponentsByKey.bind(viewer3D)
      }
    };
    this.hotKeyLner = this.hotKeyEvent.bind(this);
  }

  componentDidMount() {
    document.addEventListener("keypress", this.hotKeyLner);
  }

  componentWillUnmount() {
    document.removeEventListener("keypress", this.hotKeyLner);
  }

  hotKeyEvent(ev) {
    const pointer = this.props.pointer;
    if (ev.ctrlKey && ev.key === "z" && pointer > 0) {
      this.undoOrRedo("undo", this.props.pointer - 1, this.props.pointer);
      this.props.movePointer(this.props.pointer - 1);
    }
    if (ev.ctrlKey && ev.key === "Z" && pointer < this.props.history.length) {
      this.undoOrRedo("redo", this.props.pointer, this.props.pointer + 1);
      this.props.movePointer(this.props.pointer + 1);
    }
  }

  onClick(e, key) {
    e.preventDefault();
    e.stopPropagation();
    if (key < this.props.pointer) {
      this.undoOrRedo('undo', key, this.props.pointer);
    } else if (key > this.props.pointer) {
      this.undoOrRedo('redo', this.props.pointer, key);
    }
    this.props.movePointer(key);
  }

  /**
   * 撤销或重做i-j步
   * @desc 包括i不包括j
   * @param {"undo" | "redo"} type - 类型
   * @param {number} i
   * @param {number} j
   */
  undoOrRedo(type, i, j) {
    if (type === 'undo') {
      let hasIsolate = false;
      for (let x = j - 1; x >= i; x--) {
        const item = this.props.history[x];
        if (item.type === 'isolate') {
          hasIsolate = true;
        } else {
          this.viewerFunc['undo'][item.type](item.keys);
        }
      }
      // 撤销隔离需要找到前置操作的所有隔离和隐藏操作，然后使用操作重放
      if (hasIsolate) {
        const keyObj = {
          hide: [],
          isolate: []
        };
        for (let x = i - 1; x >= 0; x--) {
          const item = this.props.history[x];
          if (item.type === 'hide') {
            keyObj.hide.push(...item.keys);
          }
          if (item.type === 'isolate') {
            keyObj.isolate.push(...item.keys);
          }
        }
        keyObj.hide = _.uniq(keyObj.hide);
        keyObj.isolate = _.uniq(keyObj.isolate);
        this.viewerFunc['undo'].isolate(keyObj);
      }
    } else {
      for (let x = i; x < j; x++) {
        const item = this.props.history[x];
        this.viewerFunc['redo'][item.type](item.keys, item.color, item.opacity);
      }
    }
  }

  render() {
    const historyContent = [];
    const dataSource = this.props.history.length > 0 ? this.props.history : [];

    const len = dataSource.length;
    // console.log(this.props.pointer, 'this.props.pointer ')

    for (let i = 0; i < len; i++) {
      historyContent.push(
        <div
          role="button"
          tabIndex={0}
          className={`${style.item} ${i < this.props.pointer ? style.itemVisiable : style.item}`}
          key={i + 1}
          onClick={e => { this.onClick(e, i + 1) }}
        >
          <div className={style.iconWrap}>
            <AntdIcon type={i < this.props.pointer ? 'iconcomponentvisible' : 'iconconceal'} className={style.icon} />
          </div>
          <div className={style.title}>
            {dataSource[i].name}
          </div>
        </div>
      );
    }

    return (
      <Modal
        title="操作历史"
        visible={this.props.visible}
        onCancel={() => this.props.onHide()}
        viewportDiv={this.props.viewer.viewportDiv}
        top={DEFAULT_MODAL_PLACE.undo.top}
        right={DEFAULT_MODAL_PLACE.undo.right}
        left={DEFAULT_MODAL_PLACE.undo.left}

        width="350px"
        height="600px"
        minWidth={200}
        minHeight={400}
      >
        <div className={style.undoContainer}>
          <div className={style.list}>
            {historyContent.length > 0 ? historyContent : (
              <Empty>
                <div>暂无操作记录</div>
              </Empty>
            )}
          </div>
          <div className={style.footer}>
            <div className={style.line} />
            <button
              type="button"
              className="bos-btn bos-btn-primary bos-btn-block"
              onClick={e => { this.onClick(e, 0) }}
            >
              撤销列表操作
            </button>
          </div>
        </div>

      </Modal>
    );
  }
}

UndoList.propTypes = {
  pointer: PropTypes.number.isRequired,
  history: PropTypes.arrayOf(PropTypes.object).isRequired,
  movePointer: PropTypes.func.isRequired,
  viewer: PropTypes.object.isRequired,
  visible: PropTypes.bool.isRequired,
  onHide: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  pointer: state.button.pointer,
  history: state.button.history,
  viewer: state.system.viewer3D,
});
const mapDispatchToProps = (dispatch) => ({
  movePointer: pointer => { dispatch(movePointer((pointer))) }
});
const WrappedContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(UndoList);

export default WrappedContainer;
