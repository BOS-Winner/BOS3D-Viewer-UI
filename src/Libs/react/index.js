import React from "react";
import useUpdateEffect from "./useUpdateEffect";

React.useUpdateEffect = useUpdateEffect;

export default React;
// export { default as useUpdateEffect } from "./useUpdateEffect";
