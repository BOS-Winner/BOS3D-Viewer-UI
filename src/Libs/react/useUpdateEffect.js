import { useEffect, useRef } from "react";

export default function (effect, deps) {
  const isFirstRender = useRef(true);
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
    } else {
      effect();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}
