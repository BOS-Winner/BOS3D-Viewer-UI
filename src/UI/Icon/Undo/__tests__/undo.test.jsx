import React from 'react';
import { render, act } from '@testing-library/react';
import { Provider } from 'react-redux';
import { genStore } from 'jest/initTest';
import UndoList from '../UndoList';
import { updateList } from '../../action';

jest.mock('Base/Modal', () => {
  // eslint-disable-next-line global-require
  const Modal = require('mock/Modal').default;
  return props => <Modal {...props} />;
});
// store.getState().system.viewer3D.resetScene = jest.fn();
const onHide = jest.fn();
let store;
let undoContainer;
let undoGetByText;
let undoRerender;

describe('undo list', () => {
  beforeEach(() => {
    store = genStore();
    store.getState().system.viewer3D.showComponentsByKey = jest.fn();
    store.getState().system.viewer3D.hideComponentsByKey = jest.fn();
    store.getState().system.viewer3D.isolateComponentsByKey = jest.fn();
    store.getState().system.viewer3D.resetScene = jest.fn();
    store.getState().system.viewer3D.colorfulComponentsByKey = jest.fn();
    store.getState().system.viewer3D.closeColorfulComponentsByKey = jest.fn();
    store.getState().system.viewer3D.wireFrameComponentsByKey = jest.fn();
    store.getState().system.viewer3D.closeWireFrameComponentsByKey = jest.fn();
    const { container, getByText, rerender } = render((
      <Provider store={store}>
        <UndoList visible onHide={onHide} />
      </Provider>
    ));

    undoContainer = container;
    undoGetByText = getByText;
    undoRerender = rerender;
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should show or hide', () => {
    expect(undoContainer).toMatchSnapshot();
    act(() => {
      undoGetByText('close mock modal').click();
    });
    expect(onHide).toBeCalledTimes(1);
    undoRerender((
      <Provider store={store}>
        <UndoList visible={false} onHide={onHide} />
      </Provider>
    ));
    expect(undoContainer).toMatchSnapshot();
  });

  it('should add item', () => {
    store.dispatch(updateList({
      type: 'hide',
      name: "构件隐藏",
      keys: ['M123456']
    }));
    expect(undoContainer).toMatchSnapshot();
  });

  it('should undo or redo action', () => {
    const viewer3D = store.getState().system.viewer3D;
    store.dispatch(updateList({
      type: 'hide',
      name: "构件隐藏",
      keys: ['M123456']
    }));
    store.dispatch(updateList({
      type: 'wireframe',
      name: "构件线框化",
      keys: ['M1234567']
    }));
    store.dispatch(updateList({
      type: 'isolate',
      name: "构件隔离",
      keys: ['M12345678']
    }));
    store.dispatch(updateList({
      type: 'colorful',
      name: "构件变色",
      keys: ['M12345678']
    }));

    act(() => {
      undoGetByText('构件线框化').click();
    });
    expect(viewer3D.closeColorfulComponentsByKey).toBeCalledTimes(1);
    expect(viewer3D.isolateComponentsByKey).toBeCalledTimes(0);
    expect(viewer3D.resetScene).toBeCalledTimes(1);
    expect(viewer3D.hideComponentsByKey).toBeCalledTimes(1);
    expect(viewer3D.closeWireFrameComponentsByKey).toBeCalledTimes(0);
    expect(viewer3D.showComponentsByKey).toBeCalledTimes(0);
    expect(undoContainer).toMatchSnapshot();

    act(() => {
      undoGetByText('撤销列表操作').click();
    });
    expect(viewer3D.closeWireFrameComponentsByKey).toBeCalledTimes(1);
    expect(viewer3D.showComponentsByKey).toBeCalledTimes(1);
    expect(undoContainer).toMatchSnapshot();

    act(() => {
      undoGetByText('构件隔离').click();
    });
    expect(viewer3D.colorfulComponentsByKey).toBeCalledTimes(0);
    expect(viewer3D.isolateComponentsByKey).toBeCalledTimes(1);
    expect(viewer3D.wireFrameComponentsByKey).toBeCalledTimes(1);
    expect(viewer3D.hideComponentsByKey).toBeCalledTimes(2);
    expect(undoContainer).toMatchSnapshot();

    act(() => {
      undoGetByText('构件变色').click();
    });
    expect(viewer3D.colorfulComponentsByKey).toBeCalledTimes(1);
    expect(viewer3D.isolateComponentsByKey).toBeCalledTimes(1);
    expect(viewer3D.wireFrameComponentsByKey).toBeCalledTimes(1);
    expect(viewer3D.hideComponentsByKey).toBeCalledTimes(2);
    expect(undoContainer).toMatchSnapshot();
  });
});
