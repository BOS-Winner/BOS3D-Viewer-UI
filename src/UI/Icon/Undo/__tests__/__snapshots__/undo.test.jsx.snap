// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`undo list should add item 1`] = `
<div>
  <div
    style="display: block; position: absolute; top: 20%; left: 20px; width: 163px; height: 200px;"
    title="操作历史"
  >
    <div>
      <span>
        操作历史
      </span>
      <button
        data-testid="close-mock-modal"
        type="button"
      >
        close mock modal
      </button>
    </div>
    <div
      class="list"
    >
      <div
        class="item"
        role="button"
        tabindex="0"
      >
        撤销列表操作
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(133, 210, 255, 0.698); color: rgb(255, 255, 255);"
        tabindex="0"
      >
        构件隐藏
      </div>
    </div>
  </div>
</div>
`;

exports[`undo list should show or hide 1`] = `
<div>
  <div
    style="display: block; position: absolute; top: 20%; left: 20px; width: 163px; height: 200px;"
    title="操作历史"
  >
    <div>
      <span>
        操作历史
      </span>
      <button
        data-testid="close-mock-modal"
        type="button"
      >
        close mock modal
      </button>
    </div>
    <div
      class="list"
    >
      <div
        class="item"
        role="button"
        tabindex="0"
      >
        撤销列表操作
      </div>
    </div>
  </div>
</div>
`;

exports[`undo list should show or hide 2`] = `
<div>
  <div
    style="display: none; position: absolute; top: 20%; left: 20px; width: 163px; height: 200px;"
    title="操作历史"
  >
    <div>
      <span>
        操作历史
      </span>
      <button
        data-testid="close-mock-modal"
        type="button"
      >
        close mock modal
      </button>
    </div>
    <div
      class="list"
    >
      <div
        class="item"
        role="button"
        tabindex="0"
      >
        撤销列表操作
      </div>
    </div>
  </div>
</div>
`;

exports[`undo list should undo or redo action 1`] = `
<div>
  <div
    style="display: block; position: absolute; top: 20%; left: 20px; width: 163px; height: 200px;"
    title="操作历史"
  >
    <div>
      <span>
        操作历史
      </span>
      <button
        data-testid="close-mock-modal"
        type="button"
      >
        close mock modal
      </button>
    </div>
    <div
      class="list"
    >
      <div
        class="item"
        role="button"
        tabindex="0"
      >
        撤销列表操作
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(133, 210, 255, 0.698); color: rgb(255, 255, 255);"
        tabindex="0"
      >
        构件隐藏
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(133, 210, 255, 0.698); color: rgb(255, 255, 255);"
        tabindex="0"
      >
        构件线框化
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(255, 255, 255, 0.4); color: rgb(235, 235, 235);"
        tabindex="0"
      >
        构件隔离
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(255, 255, 255, 0.4); color: rgb(235, 235, 235);"
        tabindex="0"
      >
        构件变色
      </div>
    </div>
  </div>
</div>
`;

exports[`undo list should undo or redo action 2`] = `
<div>
  <div
    style="display: block; position: absolute; top: 20%; left: 20px; width: 163px; height: 200px;"
    title="操作历史"
  >
    <div>
      <span>
        操作历史
      </span>
      <button
        data-testid="close-mock-modal"
        type="button"
      >
        close mock modal
      </button>
    </div>
    <div
      class="list"
    >
      <div
        class="item"
        role="button"
        tabindex="0"
      >
        撤销列表操作
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(255, 255, 255, 0.4); color: rgb(235, 235, 235);"
        tabindex="0"
      >
        构件隐藏
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(255, 255, 255, 0.4); color: rgb(235, 235, 235);"
        tabindex="0"
      >
        构件线框化
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(255, 255, 255, 0.4); color: rgb(235, 235, 235);"
        tabindex="0"
      >
        构件隔离
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(255, 255, 255, 0.4); color: rgb(235, 235, 235);"
        tabindex="0"
      >
        构件变色
      </div>
    </div>
  </div>
</div>
`;

exports[`undo list should undo or redo action 3`] = `
<div>
  <div
    style="display: block; position: absolute; top: 20%; left: 20px; width: 163px; height: 200px;"
    title="操作历史"
  >
    <div>
      <span>
        操作历史
      </span>
      <button
        data-testid="close-mock-modal"
        type="button"
      >
        close mock modal
      </button>
    </div>
    <div
      class="list"
    >
      <div
        class="item"
        role="button"
        tabindex="0"
      >
        撤销列表操作
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(133, 210, 255, 0.698); color: rgb(255, 255, 255);"
        tabindex="0"
      >
        构件隐藏
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(133, 210, 255, 0.698); color: rgb(255, 255, 255);"
        tabindex="0"
      >
        构件线框化
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(133, 210, 255, 0.698); color: rgb(255, 255, 255);"
        tabindex="0"
      >
        构件隔离
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(255, 255, 255, 0.4); color: rgb(235, 235, 235);"
        tabindex="0"
      >
        构件变色
      </div>
    </div>
  </div>
</div>
`;

exports[`undo list should undo or redo action 4`] = `
<div>
  <div
    style="display: block; position: absolute; top: 20%; left: 20px; width: 163px; height: 200px;"
    title="操作历史"
  >
    <div>
      <span>
        操作历史
      </span>
      <button
        data-testid="close-mock-modal"
        type="button"
      >
        close mock modal
      </button>
    </div>
    <div
      class="list"
    >
      <div
        class="item"
        role="button"
        tabindex="0"
      >
        撤销列表操作
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(133, 210, 255, 0.698); color: rgb(255, 255, 255);"
        tabindex="0"
      >
        构件隐藏
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(133, 210, 255, 0.698); color: rgb(255, 255, 255);"
        tabindex="0"
      >
        构件线框化
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(133, 210, 255, 0.698); color: rgb(255, 255, 255);"
        tabindex="0"
      >
        构件隔离
      </div>
      <div
        class="item"
        role="button"
        style="background-color: rgba(133, 210, 255, 0.698); color: rgb(255, 255, 255);"
        tabindex="0"
      >
        构件变色
      </div>
    </div>
  </div>
</div>
`;
